DIPs/1NNN-ADR.md

```diff
@@ -28,25 +84,71 @@ embedding the arguments in the string itself.
 
 ## Rationale
 
-While the conventional format string followed by the argument list is fine for
-short strings and a small number of arguments, it tends to break down with longer strings
-that have many arguments. Omitted arguments, extra arguments, and mismatches
-between format specifiers and their corresponding arguments are common errors. Embedding arguments
-in the format strings can eliminate these errors. Readability is improved and the code is visually
-easier to review for correctness.
+This proposal was born from the rejection of DIP1027 that had the desire to have interpolated strings 
+found in other languages and to have a generic capability to use the interpolated syntax with other domains.  For example,
+it should be possible to have an SQL aware transform.
+
+```D
+    --- Put compelling SQL example here... ---
+    auto query = "SELECT field1, field2 FROM table WHERE field1 = :field1Value;";
+    auto stmt = db.prepare( query );
+    stmt.bind( ":field1Value", value1 );
+```
+
+Another example is the generation of HTML tags
+
+```D
+    --- Put compelling HTML example here ---
+    auto string div = "<div id=\"" ~ id ~ "\">"
+```
+
+These examples both benefit from the proposed interpolation syntax because the compiler can
+semantically verify the correctness of the expression while preparing a tuple suitable for arguments
+for subsequent function calls.
+
+```D
+    auto query = i"SELECT field1, field2 FROM table WHERE field1 = ${field1Value};";
+    auto string div = "<div id=\"${id}\">";
```

Comment:
You need .idup here for our proposal.

Reply:
In general, I wonder if the html example is compelling. In many many cases, just passing the items in the order needed works just fine. Where it's really useful is in 3 places:

1. you want to just generate a string from other data. This is the `idup` case.
2. You want to interface with existing functions like printf/writeln/db query.
3. You have a specific need to separate the "format" from the "data". An example here is SQL injection attacks -- you want to know which parts of the SQL are structure, and which parts are data.

This DIP gives you all these things, and has the base to be able to avoid even parsing the format string. The more I think about this, the more I think we should embrace the fact that you can avoid having to reparse the format string for case 3. It suddenly is very annoying for me that writefln(i"...") is going to forward itself to the stupid writefln that is then going to reparse that which was already parsed by the compiler. I know it's something we can plan on for the future, but maybe we just put that out there right now.